File: dp_primer/dp_primer_partial_sums_step2.cs
// 実行時間 20ms
using static System.Console;
using static System.Linq.Enumerable;

internal class Program
{
    static void Main()
    {
        const int b = 200;
        int[] nx = ReadLine().Split( ' ' ).Select( int.Parse ).ToArray();
        int[] a = new int[nx[0]];
        int[] p = Repeat( b, nx[1] + 1 ).ToArray();

        foreach( int i in Range( 0, nx[0] ) ) {
            a[i] = int.Parse( ReadLine() );
        }
        p[0] = 0;
        foreach( int i in Range( 0, nx[0] ) ) {
            for( int j = nx[1]; a[i] <= j; j-- ) {
                p[j] = System.Math.Min( p[j], p[j - a[i]] + 1 );
            }
        }
        WriteLine( p[nx[1]] != b ? p[nx[1]] : -1 );
        return;
    }
}

File: dp_primer/dp_primer_partial_sums_step2.py
# 実行時間 30ms
B = 200
n, x = list(map(int, input().split()))
a = []
p = [B for _ in [0] * (x + 1)]

for _ in range(n):
    a.append(int(input()))
p[0] = 0
for i in range(n):
    for j in range(x, a[i] - 1, -1):
        if p[j - a[i]] + 1 < p[j]:
            p[j] = p[j - a[i]] + 1
print(p[x] if p[x] != B else -1)
